File: src/air_mouse_mission/air_mouse_mission/map_generator.py
#!/usr/bin/env python3
"""
map_generator.py - Generates final 2D map for NIDAR AirMouse scoring
Requirements: Generate and display a 2D map of explored area

Input:
- /map (OccupancyGrid) from mapping_node / slam_toolbox
- /scan_3d_world (PointCloud2) from nodding lidar
- /survivor_markers

Output:
- /tmp/nidar_2d_map.pgm + .yaml (ROS map format)
- /tmp/nidar_2d_map_with_survivors.png (visual for judges)
- Publishes /final_2d_map for ground station

NIDAR scoring: Map quality, coverage, survivor tags
"""
import rclpy
from rclpy.node import Node
from nav_msgs.msg import OccupancyGrid
from visualization_msgs.msg import MarkerArray
import numpy as np
import cv2
import yaml
import os

class MapGenerator(Node):
    def __init__(self):
        super().__init__('map_generator')
        self.map = None
        self.map_info = None
        self.survivors = []

        self.map_sub = self.create_subscription(OccupancyGrid, '/map', self.map_callback, 10)
        self.survivor_sub = self.create_subscription(MarkerArray, '/survivor_markers', self.survivor_callback, 10)

        self.timer = self.create_timer(5.0, self.generate)

        self.get_logger().info("NIDAR Map Generator ready")

    def map_callback(self, msg):
        self.map = np.array(msg.data, dtype=np.int8).reshape((msg.info.height, msg.info.width))
        self.map_info = msg.info

    def survivor_callback(self, msg):
        self.survivors = []
        for m in msg.markers:
            if m.ns == "survivors":
                self.survivors.append((m.pose.position.x, m.pose.position.y, m.id))

    def generate(self):
        if self.map is None:
            return

        # Create visual map
        # ROS map: -1 unknown, 0 free, 100 occupied
        # Image: unknown gray 128, free white 255, occupied black 0
        img = np.zeros((self.map.shape[0], self.map.shape[1], 3), dtype=np.uint8)
        img[self.map == -1] = [128, 128, 128]
        img[self.map == 0] = [255, 255, 255]
        img[self.map == 100] = [0, 0, 0]

        # Draw survivors as orange circles with ID
        if self.map_info is not None:
            for x, y, sid in self.survivors:
                gx = int((x - self.map_info.origin.position.x) / self.map_info.resolution)
                gy = int((y - self.map_info.origin.position.y) / self.map_info.resolution)
                if 0 <= gx < img.shape[1] and 0 <= gy < img.shape[0]:
                    cv2.circle(img, (gx, gy), 8, (0, 165, 255), -1)  # orange BGR
                    cv2.putText(img, f"S{sid+1}", (gx+10, gy), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0,0,255), 2)

            # Draw launch pad at home (0,0) - 2ft x 2ft = 0.61m
            home_gx = int((0 - self.map_info.origin.position.x) / self.map_info.resolution)
            home_gy = int((0 - self.map_info.origin.position.y) / self.map_info.resolution)
            size_px = int(0.61 / self.map_info.resolution)
            cv2.rectangle(img, (home_gx-size_px//2, home_gy-size_px//2), (home_gx+size_px//2, home_gy+size_px//2), (0,255,0), 2)
            cv2.putText(img, "HOME", (home_gx-20, home_gy-10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0,255,0), 2)

        # Flip for correct orientation (ROS origin bottom-left, image top-left)
        img_flipped = cv2.flip(img, 0)

        # Save PNG for judges
        cv2.imwrite('/tmp/nidar_2d_map_with_survivors.png', img_flipped)
        cv2.imwrite('/tmp/nidar_2d_map.png', img_flipped)

        # Save PGM + YAML for ROS nav
        pgm_path = '/tmp/nidar_2d_map.pgm'
        yaml_path = '/tmp/nidar_2d_map.yaml'

        # PGM: grayscale
        gray = cv2.cvtColor(img_flipped, cv2.COLOR_BGR2GRAY)
        # Map values: free 254, occupied 0, unknown 205
        pgm_data = np.zeros_like(gray, dtype=np.uint8)
        pgm_data[self.map == 0] = 254
        pgm_data[self.map == 100] = 0
        pgm_data[self.map == -1] = 205
        pgm_data = np.flipud(pgm_data)  # PGM origin bottom-left

        cv2.imwrite(pgm_path, pgm_data)

        # YAML
        yaml_data = {
            'image': pgm_path,
            'resolution': float(self.map_info.resolution),
            'origin': [float(self.map_info.origin.position.x), float(self.map_info.origin.position.y), 0.0],
            'negate': 0,
            'occupied_thresh': 0.65,
            'free_thresh': 0.196
        }
        with open(yaml_path, 'w') as f:
            yaml.dump(yaml_data, f)

        self.get_logger().info(f"NIDAR 2D map saved: {pgm_path}, {yaml_path}, PNG with {len(self.survivors)} survivors")

def main(args=None):
    rclpy.init(args=args)
    node = MapGenerator()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/air_mouse_cpp/src/map_generator.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/occupancy_grid.hpp>
#include <visualization_msgs/msg/marker_array.hpp>
#include <opencv2/opencv.hpp>
#include <yaml-cpp/yaml.h>
#include <vector>
#include <fstream>

class MapGenerator : public rclcpp::Node {
public:
  MapGenerator() : Node("map_generator") {
    map_sub_=this->create_subscription<nav_msgs::msg::OccupancyGrid>("/map", 10, std::bind(&MapGenerator::mapCb, this, std::placeholders::_1));
    surv_sub_=this->create_subscription<visualization_msgs::msg::MarkerArray>("/survivor_markers", 10, std::bind(&MapGenerator::survCb, this, std::placeholders::_1));
    timer_=this->create_wall_timer(std::chrono::seconds(5), std::bind(&MapGenerator::generate, this));
    RCLCPP_INFO(this->get_logger(), "C++ Map Generator");
  }

private:
  void mapCb(const nav_msgs::msg::OccupancyGrid::SharedPtr msg) { map_=*msg; has_map_=true; }
  void survCb(const visualization_msgs::msg::MarkerArray::SharedPtr msg) {
    survivors_.clear();
    for (auto &m: msg->markers) if (m.ns=="survivors") survivors_.push_back(m);
  }

  void generate() {
    if (!has_map_) return;
    int w=map_.info.width, h=map_.info.height;
    cv::Mat img(h, w, CV_8UC3, cv::Scalar(128,128,128));
    for (int y=0; y<h; ++y) {
      for (int x=0; x<w; ++x) {
        int idx=y*w+x;
        int8_t v=map_.data[idx];
        if (v==0) img.at<cv::Vec3b>(y,x)=cv::Vec3b(255,255,255);
        else if (v==100) img.at<cv::Vec3b>(y,x)=cv::Vec3b(0,0,0);
      }
    }
    // Draw survivors
    for (auto &m: survivors_) {
      int gx = static_cast<int>((m.pose.position.x - map_.info.origin.position.x)/map_.info.resolution);
      int gy = static_cast<int>((m.pose.position.y - map_.info.origin.position.y)/map_.info.resolution);
      if (gx>=0&&gx<w&&gy>=0&&gy<h) {
        cv::circle(img, cv::Point(gx,gy), 8, cv::Scalar(0,165,255), -1);
        cv::putText(img, "S"+std::to_string(m.id+1), cv::Point(gx+10,gy), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0,0,255),2);
      }
    }
    // Home
    int home_gx = static_cast<int>((0 - map_.info.origin.position.x)/map_.info.resolution);
    int home_gy = static_cast<int>((0 - map_.info.origin.position.y)/map_.info.resolution);
    int size_px = static_cast<int>(0.61/map_.info.resolution);
    cv::rectangle(img, cv::Point(home_gx-size_px/2, home_gy-size_px/2), cv::Point(home_gx+size_px/2, home_gy+size_px/2), cv::Scalar(0,255,0),2);

    cv::Mat flipped; cv::flip(img, flipped, 0);
    cv::imwrite("/tmp/nidar_2d_map_with_survivors.png", flipped);
    cv::imwrite("/tmp/nidar_2d_map.png", flipped);

    // PGM
    cv::Mat gray; cv::cvtColor(flipped, gray, cv::COLOR_BGR2GRAY);
    cv::Mat pgm(h,w,CV_8UC1);
    for (int y=0;y<h;++y) for (int x=0;x<w;++x) {
      int idx=(h-1-y)*w+x; // flip back for pgm
      int8_t v=map_.data[idx];
      if (v==0) pgm.at<uchar>(y,x)=254;
      else if (v==100) pgm.at<uchar>(y,x)=0;
      else pgm.at<uchar>(y,x)=205;
    }
    cv::imwrite("/tmp/nidar_2d_map.pgm", pgm);

    YAML::Emitter out;
    out << YAML::BeginMap;
    out << YAML::Key << "image" << YAML::Value << "/tmp/nidar_2d_map.pgm";
    out << YAML::Key << "resolution" << YAML::Value << map_.info.resolution;
    out << YAML::Key << "origin" << YAML::Value << YAML::Flow << YAML::BeginSeq << map_.info.origin.position.x << map_.info.origin.position.y << 0.0 << YAML::EndSeq;
    out << YAML::Key << "negate" << YAML::Value << 0;
    out << YAML::Key << "occupied_thresh" << YAML::Value << 0.65;
    out << YAML::Key << "free_thresh" << YAML::Value << 0.196;
    out << YAML::EndMap;
    std::ofstream yaml("/tmp/nidar_2d_map.yaml");
    yaml<<out.c_str();

    RCLCPP_INFO(this->get_logger(), "Map saved with %zu survivors", survivors_.size());
  }

  nav_msgs::msg::OccupancyGrid map_;
  bool has_map_=false;
  std::vector<visualization_msgs::msg::Marker> survivors_;
  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_sub_;
  rclcpp::Subscription<visualization_msgs::msg::MarkerArray>::SharedPtr surv_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MapGenerator>());
  rclcpp::shutdown();
  return 0;
}
